contracts/token/psp721/traits.rs

```diff
@@ -18,62 +52,236 @@ pub enum PSP721Error {
+    }
 
     /// Approves the account to transfer the specified token on behalf of the caller.
     #[ink(message)]
-    fn approve(&mut self, to: AccountId, id: Id);
+    fn approve(&mut self, to: AccountId, id: Id) {
+        self._approve_for(to, id);
+    }
 
     /// Transfer approved or owned token.
     #[ink(message)]
-    fn transfer_from(&mut self, from: AccountId, to: AccountId, id: Id);
+    fn transfer_from(&mut self, from: AccountId, to: AccountId, id: Id) {
+        self._transfer_token_from(&from, to.clone(), id);
+        self._emit_transfer_event(from, to, id);
+    }
 
     /// Transfers token with `id` from `from` to `to`. Also some `data` can be passed.
     #[ink(message)]
-    fn safe_transfer_from(&mut self, from: AccountId, to: AccountId, id: Id, data: Vec<u8>);
+    fn safe_transfer_from(&mut self, from: AccountId, to: AccountId, id: Id, data: Vec<u8>) {
+        self._transfer_token_from(&from, to.clone(), id);
+        self._call_contract_transfer(Self::env().caller(), from, to, id, data);
+        self._emit_transfer_event(from, to, id);
+    }
+
+    // Helper functions
+
+    /// Emits transfer event. This method must be implemented in derived implementation
+    fn _emit_transfer_event(&self, _from: AccountId, _to: AccountId, _id: Id) {
+        // TODO: Emit events
+    }
+
+    /// Emits approval event. This method must be implemented in derived implementation
+    fn _emit_approval_event(&self, _from: AccountId, _to: AccountId, _id: Id) {
+        // TODO: Emit events
+    }
+
+    /// Emits approval for all event. This method must be implemented in derived implementation
+    fn _emit_approval_for_all_event(&self, _owner: AccountId, _operator: AccountId, _approved: bool) {
+        // TODO: Emit events
+    }
+
+    /// Approves or disapproves the operator to transfer all tokens of the caller.
+    fn _approve_for_all(&mut self, to: AccountId, approved: bool) {
+        let caller = Self::env().caller();
+        assert_ne!(to, caller, "{}", PSP721Error::NotAllowed.as_ref());
+        self._emit_approval_for_all_event(caller, to, approved);
+        if self._approved_for_all(caller, to) {
+            let status = self
+                ._operator_approvals_mut()
+                .get_mut(&(caller, to)).unwrap();
+            *status = approved;
+        } else {
+            self._operator_approvals_mut().insert((caller, to), approved);
+        }
+    }
+
+    /// Approve the passed AccountId to transfer the specified token on behalf of the message's sender.
+    fn _approve_for(&mut self, to: AccountId, id: Id) {
+        let caller = Self::env().caller();
+        let owner = self._owner_of(&id);
+        if !(owner == Some(caller)
+            || self._approved_for_all(owner.expect("PSP721Error with AccountId"), caller))
+        {
+            panic!("{}", PSP721Error::NotAllowed.as_ref());
+        };
+
+        assert_ne!(to, [0; 32].into(), "{}", PSP721Error::NotAllowed.as_ref());
```

Comment:
maybe create a `pub(crate) const ZERO_ADDRESS: [u8; 32] = [0u8; 32]` and reuse?

Reply:
No, we can't do that, because it must be public then(After this change whole implementation of some contracts must be public because we copy implementation from one crate to another). 
We can try to use `ZERO_ADDRESS`, but I don't know how it will resolve conflicts with `ZERO_ADDRESS` in another crates=)

Reply:
Will using this const as `crate::ZERO_ADDRESS` work fine?  I guess then won't conflict with other `ZERO_ADDRESS`es